gateway/src/cron.rs

```diff
@@ -0,0 +1,427 @@
+        let total_validators = checkpoint.validators.validators().len();
+
+        self.update_submitters(store, submitter)?;
+
+        let checkpoint_hash = self.insert_checkpoint(store, checkpoint)?;
+        let most_submitted_count = self.update_submission_count(store, checkpoint_hash)?;
+
+        // use u16 numerator and denominator to avoid floating point calculation and external crate
+        // total validators should be within u16::MAX.
+        if total_validators as u16 * RATIO_NUMERATOR / RATIO_DENOMINATOR > most_submitted_count {
+            return Ok(false);
+        }
+
+        Ok(true)
+    }
+
+    pub fn load_most_submitted_checkpoint<BS: Blockstore>(
+        &self,
+        store: &BS,
+    ) -> anyhow::Result<Option<CronCheckpoint>> {
+        if let Some(most_submitted_hashes) = &self.most_submitted_hashes {
+            // we will only have one entry in the `most_submitted` set if more than 2/3 has reached
+            let hash = most_submitted_hashes.iter().next().unwrap();
+            self.get_submission(store, hash)
+        } else {
+            Ok(None)
+        }
+    }
+
+    pub fn get_submission<BS: Blockstore>(
+        &self,
+        store: &BS,
+        hash: &HashOutput,
+    ) -> anyhow::Result<Option<CronCheckpoint>> {
+        let hamt = self.submissions.load(store)?;
+        let key = BytesKey::from(hash.as_slice());
+        Ok(hamt.get(&key)?.cloned())
+    }
+
+    /// Update the total submitters, returns the latest total number of submitters
+    fn update_submitters<BS: Blockstore>(
+        &mut self,
+        store: &BS,
+        submitter: Address,
+    ) -> anyhow::Result<()> {
+        let addr_byte_key = BytesKey::from(submitter.to_bytes());
+        self.submitters.modify(store, |hamt| {
+            // check the submitter has not submitted before
+            if hamt.contains_key(&addr_byte_key)? {
+                return Err(anyhow!("already submitted"));
+            }
+
+            // now the submitter has not submitted before, mark as submitted
+            hamt.set(addr_byte_key, ())?;
+
+            Ok(())
+        })
+    }
+
+    /// Insert the checkpoint to store if it has not been submitted before. Returns the hash of the checkpoint.
+    fn insert_checkpoint<BS: Blockstore>(
+        &mut self,
+        store: &BS,
+        checkpoint: CronCheckpoint,
+    ) -> anyhow::Result<HashOutput> {
+        let hash = checkpoint.hash()?;
+        let hash_key = BytesKey::from(hash.as_slice());
+
+        let hamt = self.submissions.load(store)?;
+        if hamt.contains_key(&hash_key)? {
+            return Ok(hash);
+        }
+
+        // checkpoint has not submitted before
+        self.submissions.modify(store, |hamt| {
+            hamt.set(hash_key, checkpoint)?;
+            Ok(())
+        })?;
+
+        Ok(hash)
+    }
+
+    /// Update submission count of the hash. Returns the currently most submitted submission count.
+    fn update_submission_count<BS: Blockstore>(
```

Comment:
As mentioned above, I think we can simplify this logic using a simpler data structure. In the end we just want to check: 
- The number of votes for a submission.
- And that a submitter hasn't voted twice.

Reply:
resolving this as we have discussed over slack, see above comment